### DISCUSSION TITLE
Analog mic + pwm out behaves weird?

### DISCUSSION THREAD

**Discussion by thegoodhen** [pschatzmann/arduino-audio-tools#1480] [Q&A] [answered]
Hello, I am putting this into Q and A, since I am not sure if it's a bug or if I am just doing something wrong.

Thank you so much for this awesome library.
I started playing with it and I'm trying to build a simple Dalek voice changer.
As a first step, I am simply trying to copy the microphone data to a speaker. 

I tried reading the analog input and streaming it as PWM, but that doesn't work as expected.

I made sure I am using the correct pin for the analog input by running the base_adc_average_mono_serial example, which works.

I made sure I am using the correct pin for the analog output by running the streams_generator_pwm example, which works.

In order to have repeatable data, instead of microphone I am using a signal generator with a DC offset. I verified that the adc sees the data correctly with the  base_adc_average_mono_serial.

There is low volume noise coming from the speaker. The oscilloscope shows a waveform with a roughly similar shape to the one I am putting in, extremely noisy, low amplitude, chopped up (discontinuities in the waveform, as if you cut pieces of it and stitched them back together) and at a much lower frequency, suggesting a sampling issue.

Am I just doing something really dumb? Or is this combination not supported? Or is it a bug?

I just need a combination of an analog input and "analog-compatible" (i.e. lm386 with an aliasing filter compatible) output.

As far as I remember, the ESP32 does not support ADC DMA with I2S PDM, since ADC DMA uses the I2S0 memory and furthermore PDM is only supported by I2S0 - that's why I tried my luck with the PWM instead.

I tried enabling "auto_center_read", didnt help.
I tried changing the sample frequency between 8kHz and 44100Hz, didn't help.
I tried changing the input resolution. 8bit (which you say is not supported by most the classes) makes the noise much louder.

Thank you very much!! If you need me to do some more tests, just let me know

HW+SW:

Arduino 1.8.19, ESP32 2.0.9, NodeMCU32S

Sketch:

```
/**
   @file streams-generator-pwm.ino
   @author Phil Schatzmann
   @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-generator-pwm/README.md
   @author Phil Schatzmann
   @copyright GPLv3
*/

#include "AudioTools.h"
#define SAMPLE_RATE 16000


PWMAudioOutput pwm;
AudioInfo info(SAMPLE_RATE, 1, 16);


// On board analog to digital converter
AnalogAudioStream analog_in;
StreamCopy copier(pwm, analog_in);    // copy in to out

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);


  auto adcConfig = analog_in.defaultConfig(RX_MODE);
  adcConfig.copyFrom(info);


  // adcConfig.is_auto_center_read = true;


  analog_in.begin(adcConfig);


  // setup PWM output
  auto config = pwm.defaultConfig();
  config.copyFrom(info);
  config.resolution = 8;  // must be between 8 and 11 -> drives pwm frequency (8 is default)

  Pins pins = {23};
  config.setPins(pins);
  pwm.begin(config);
}

void loop() {
  copier.copy();
  //delay(10);
}
```

**Comment by pschatzmann** [comment, marked as answer]
I am not sure what the issue is, but maybe there is a difference between the input and output speed which leads to a buffer over or underflow. 
- You can try to measure the input and output speed and compare them.
- You can try to play with the buffer size and count configuration both on the input and output side

**Reply by thegoodhen** [reply]
I tried playing with the buffer size, but that didn't help. I measured the fps of the pwm, which was slightly under the target for 16kHz and more, but was correct for 8K. I am not sure how to measure the actual sample rate of the adc... any ideas?

**Reply by pschatzmann** [reply]
Have a look at the different examples using MeasuringStream, but I think you can assume that this matches with what you have requested.

**Reply by thegoodhen** [reply]
Thank you, I will try it tomorrow. So far it looks like the playback is 5 times slower than the sampling with mangling going on. I ran an FFT on the PWM filter after a RC filter and the dominant frequency is almost always 5 times smaller than what I am putting in... Hopefully  we can fix this!

**Reply by pschatzmann** [reply]
Just as a side note: the best audio quality (both for input and output) you would get with a separate audio chip (e.g. es8388) using I2S.

**Reply by thegoodhen** [reply]
Hello, yes, I am aware. My application actually does not allow this easily. The voice changer should be a part of many examples of an educational kit I am making and I need to also later perform sonification of various analog input sources, such as a high-frequency light sensor or an inductive sensor (for listening to the intereference various electronic devices make). Since I want the outputs to be Grove compatible, I can only use 2 pins to connect the speaker. Of course, I guess I could add a special output port for just the speaker, or use 2 cables to connect the speaker, but this does not seem like a very elegant solution. 

Oh well, I guess at least I will be giving something back to the open source community, since I will have to fix this thing one day.

Should I file the issues I mentioned in my other posts?

**Reply by pschatzmann** [reply]
I am not sure, after all you confirm that the output of a sine wave seems to work fine and I assume you tested the input with a CSV sketch.

The issue with the timer has already been corrected.

**Reply by thegoodhen** [reply]
When you are saying the issue with the timer has been corrected, do you mean that you fixed PWM not working in Arduino 3.0 ? Also thank you very much, you are extremely supportive. Is this your fulltime job or something? It's crazy to see such a well maintained open source project.

I did test the input with a CSV sketch, yes. That's what's confusing, both the input and output seem to work fine, but when you combine them, it stops working.

**Reply by thegoodhen** [reply]
> Have a look at the different examples using MeasuringStream, but I think you can assume that this matches with what you have requested.

This does not seem to be the case. Here are some inputs of the base_adc_measure.ino sketch!


```
[I] AnalogConfigESP32.h : 50 - I2S_MODE_ADC_BUILT_IN
Starting ADC...
[I] AnalogAudio.h : 64 - bool audio_tools::AnalogAudioStream::begin(audio_tools::AnalogConfig)
[I] AnalogAudioESP32.h : 47 - virtual bool audio_tools::AnalogDriverESP32::begin(audio_tools::AnalogConfigESP32)
[I] AudioTypes.h : 124 - sample_rate: 8000
[I] AudioTypes.h : 125 - channels: 1
[I] AudioTypes.h : 126 - bits_per_sample: 16
[I] AnalogAudioESP32.h : 51 - auto_center
[I] AnalogAudioESP32.h : 94 - RX_MODE
Starting through put measurement...
==> Bytes per second: 69000
==> Bytes per second: 87000
==> Bytes per second: 87000
==> Bytes per second: 87000
==> Bytes per second: 87000
==> Bytes per second: 87000
==> Bytes per second: 87000
```
```

[I] AnalogConfigESP32.h : 50 - I2S_MODE_ADC_BUILT_IN
Starting ADC...
[I] AnalogAudio.h : 64 - bool audio_tools::AnalogAudioStream::begin(audio_tools::AnalogConfig)
[I] AnalogAudioESP32.h : 47 - virtual bool audio_tools::AnalogDriverESP32::begin(audio_tools::AnalogConfigESP32)
[I] AudioTypes.h : 124 - sample_rate: 16000
[I] AudioTypes.h : 125 - channels: 1
[I] AudioTypes.h : 126 - bits_per_sample: 16
[I] AnalogAudioESP32.h : 51 - auto_center
[I] AnalogAudioESP32.h : 94 - RX_MODE
Starting through put measurement...
==> Bytes per second: 115000
==> Bytes per second: 175000
==> Bytes per second: 175000
==> Bytes per second: 175000
==> Bytes per second: 175000

```
```
[I] AnalogConfigESP32.h : 50 - I2S_MODE_ADC_BUILT_IN
Starting ADC...
[I] AnalogAudio.h : 64 - bool audio_tools::AnalogAudioStream::begin(audio_tools::AnalogConfig)
[I] AnalogAudioESP32.h : 47 - virtual bool audio_tools::AnalogDriverESP32::begin(audio_tools::AnalogConfigESP32)
[I] AudioTypes.h : 124 - sample_rate: 44100
[I] AudioTypes.h : 125 - channels: 1
[I] AudioTypes.h : 126 - bits_per_sample: 16
[I] AnalogAudioESP32.h : 51 - auto_center
[I] AnalogAudioESP32.h : 94 - RX_MODE
Starting through put measurement...
==> Bytes per second: 69000
==> Bytes per second: 87000
==> Bytes per second: 87000
==> Bytes per second: 87000

```

Should I submit a bug report? I can also try to fix it myself and then submit a pull request.

**Reply by pschatzmann** [reply]
It would be great to file a bug report and if you manage to submit some corrections it would be greatly appreciated. 
These numbers look really strange!

**Reply by thegoodhen** [reply]
Okay! I do not see any problem so far though! The code simply seems to pass the numbers directly to i2s_driver_install... I have to investigate. I will file a bug report tomorrow evening after I have sufficient information and maybe a fix if everything goes well!

**Reply by thegoodhen** [reply]
I am starting to think it's an IDF issue.
I tried printing out  Serial.println(i2s_get_clk((i2s_port_t)0)); and it gives the sample rate I set, but the actual sample rate seems to be different.
https://github.com/espressif/esp-idf/issues/8874
https://github.com/espressif/esp-idf/issues/6691

**Reply by pschatzmann** [reply]
I agree, but setting the apll should help:

adcConfig.use_apll = true;

**Reply by thegoodhen** [reply]
Hmm, so far that only prevented the measuring stream from printing out anything, but it's late now, so maybe I am just messing something up. I will let you know!

**Comment by thegoodhen** [comment]
Okay... thank you. So the code looks good? 
I tried switching to the AudioESP32ULP output, but unfortunately that gives pretty much similar results.
Also... The ULP thing does not compile, the function bool end() seems to override void end() in AudioOutput (incorrect return type). 
When I fixed this, the AudioESP32ULP produces a sawtooth when I try to feed a sine in... Not sure why!

I will try to measure the input and output sample rates later today, I will also upgrade to the newest ESP IDF and also try to change the buffer size config. Thank you very much!

**Reply by thegoodhen** [reply]
So apparently, the example code for PWM produces silence with Arduino core 3.alpha, even though the same code works for the earlier version. Should I create an issue?

**Reply by pschatzmann** [reply]
Setting the apll might help:

  adcConfig.use_apll = true;
